src/core/tasks_pool.c

```diff
@@ -0,0 +1,137 @@
+#include "tasks_pool.h"
+
+xccl_status_t tasks_pool_init(xccl_progress_queue_t* handle) {
+    handle->ctx = (void*) malloc(sizeof(xccl_tasks_pool_t));
+    xccl_tasks_pool_t* ctx = (xccl_tasks_pool_t*) handle->ctx;
+    ctx->deque_size = (xccl_lib_global_config.tasks_pool_size / LINE_SIZE) + 1;
+    ctx->tasks = (ucc_coll_task_t ***) calloc(2 * ctx->deque_size, sizeof(ucc_coll_task_t **));
+    if (ctx->tasks == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ctx->tasks[0] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+    if (ctx->tasks[0] == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ctx->tasks[ctx->deque_size] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+    if (ctx->tasks[ctx->deque_size] == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ucs_spinlock_init(&(ctx->increase_pool_locks[0]), 0);
+    ucs_spinlock_init(&(ctx->increase_pool_locks[1]), 0);
+    ucs_spinlock_init(&(ctx->linked_list_lock), 0);
+    ctx->linked_list_head              = NULL;
+    ctx->linked_list_tail              = NULL;
+    ctx->which_pool                    = 0;
+    handle->api.progress_queue_enqueue = &tasks_pool_insert;
+    handle->api.progress_queue_dequeue = &tasks_pool_pop;
+    handle->api.progress_queue_cleanup = &tasks_pool_cleanup;
+    return XCCL_OK;
+}
+
+xccl_status_t increase_tasks(xccl_tasks_pool_t *ctx, int i) {
+    ucs_spin_lock(&ctx->increase_pool_locks[i / ctx->deque_size]);
+    if (ctx->tasks[i] == 0) {
+        ctx->tasks[i] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+        if (ctx->tasks[i] == NULL) {
+            return XCCL_ERR_NO_MEMORY;
+        }
+    }
+    ucs_spin_unlock(&ctx->increase_pool_locks[i / ctx->deque_size]);
+    return XCCL_OK;
+}
+
+xccl_status_t tasks_pool_insert(xccl_progress_queue_t *handle, ucc_coll_task_t *task) {
+    xccl_tasks_pool_t *ctx = (xccl_tasks_pool_t*) handle->ctx;
+    int i, j;
+    int which_pool = task->was_progressed ^(ctx->which_pool & 1);
+    int iteration_start = which_pool * ctx->deque_size;
+    for (i = iteration_start; i < (iteration_start + ctx->deque_size); i++) {
+        if (ctx->tasks[i] == 0) {
+            xccl_status_t status = increase_tasks(ctx, i);
+            if (status != XCCL_OK) {
+                return status;
+            }
+            return tasks_pool_insert(handle, task);
+        }
+        for (j = 0; j < LINE_SIZE; j++) {
+            /* TODO: Change atomics to UCS once release v1.9.1 is out */
+            if (__sync_bool_compare_and_swap(&(ctx->tasks[i][j]), 0, task)) {
+                return XCCL_OK;
+            }
+        }
+    }
+    ucs_spin_lock(&ctx->linked_list_lock);
+    if (ctx->linked_list_head == NULL) {
```

Comment:
all this will just be ucs_list_add_tail(list, object)

Reply:
fixed